environs/context/callcontext.go

```diff
@@ -7,51 +7,67 @@ import (
+	InvalidateModelCredentialFunc func(reason string) error
+
+	// InvalidateModelCredentialFuncGetter returns a function which records a credential as being invalid.
+	InvalidateModelCredentialFuncGetter func() (InvalidateModelCredentialFunc, error)
+
+	// CredentialIDFuncGetter returns a function which returns a credential ID.
+	CredentialIDFuncGetter func() (credential.ID, error)
+
+	// InvalidateCredentialFunc records a credential with the given ID as being invalid.
+	InvalidateCredentialFunc func(ctx stdcontext.Context, id credential.ID, reason string) error
+)
+
 // ModelCredentialInvalidator defines a point of use interface for invalidating
 // a model credential.
 type ModelCredentialInvalidator interface {
 	// InvalidateModelCredential invalidate cloud credential for the model.
 	InvalidateModelCredential(string) error
 }
 
-// NewCredentialInvalidator returns an instance which is used by providers to
-// mark a credential as invalid.
+// NewCredentialInvalidator creates a credential validator with
+// callbacks which update dqlite and mongo.
 func NewCredentialInvalidator(
-	tag names.CloudCredentialTag,
-	invalidateFunc func(ctx stdcontext.Context, id credential.ID, reason string) error,
-	legacyInvalidate func(reason string) error,
+	idGetter CredentialIDFuncGetter,
+	invalidateFunc InvalidateCredentialFunc,
+	legacyInvalidateFunc InvalidateModelCredentialFunc,
 ) ModelCredentialInvalidator {
 	return &legacyCredentialAdaptor{
-		tag:              tag,
+		idGetter:         idGetter,
 		invalidateFunc:   invalidateFunc,
-		legacyInvalidate: legacyInvalidate,
+		legacyInvalidate: legacyInvalidateFunc,
 	}
 }
 
 // legacyCredentialAdaptor exists as a *short term* solution to the fact that details
 // for credential validity exists in both dqlite (on credential records) and mongo (on models).
 // The provider calls a single InvalidateModelCredential function which updates both places.
-// TODO(wallyworld) - the legacy invalidate function needs to be changed to take a context and tag.
 type legacyCredentialAdaptor struct {
-	tag              names.CloudCredentialTag
+	idGetter         func() (credential.ID, error)
 	invalidateFunc   func(ctx stdcontext.Context, id credential.ID, reason string) error
 	legacyInvalidate func(string) error
 }
 
 // InvalidateModelCredential implements ModelCredentialInvalidator.
 func (a *legacyCredentialAdaptor) InvalidateModelCredential(reason string) error {
-	if a.tag.IsZero() {
-		return errors.New("missing credential tag")
+	credId, err := a.idGetter()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if credId.IsZero() {
+		return nil
 	}
 	if a.invalidateFunc == nil || a.legacyInvalidate == nil {
 		return errors.New("missing validation functions")
 	}
-	if err := a.invalidateFunc(stdcontext.Background(), credential.IdFromTag(a.tag), reason); err != nil {
+	if err := a.invalidateFunc(stdcontext.Background(), credId, reason); err != nil {
```

Comment:
why do we use background context here? if we ever add tracing, we will lose that knowledge here.

Reply:
This is a mistake, will fix.